ocaml/parsing/parser.mly

```diff
@@ -634,14 +634,21 @@ let pat_of_label lbl =
   Pat.mk ~loc:lbl.loc  (Ppat_var (loc_last lbl))
 
 let mk_newtypes ~loc newtypes exp =
-  let mk_one (name, jkind) exp =
+  let mk_one loc (name, jkind) exp =
     match jkind with
-    | None -> mkexp ~loc (Pexp_newtype (name, exp))
+    | None -> Exp.mk ~loc (Pexp_newtype (name, exp))
     | Some jkind ->
-      Jane_syntax.Layouts.expr_of ~loc:(make_loc loc)
+      Jane_syntax.Layouts.expr_of ~loc
         (Lexp_newtype (name, jkind, exp))
   in
-  List.fold_right mk_one newtypes exp
+  (* This is doing a fold right while making sure only
+     the outermost expression has non-ghost location *)
+  let rec inner acc = function
+    | x :: ((_ :: _) as xs) -> inner (mk_one (ghost_loc loc) x acc) xs
+    | [x] -> mk_one (make_loc loc) x acc
+    | [] -> acc
+  in
+  inner exp (List.rev newtypes)
```

Comment:
Perhaps it would be simpler to construct all nested nodes with a ghost expression, and then set the location on the outermost node to non-ghost.

Reply:
Yeah, I was hoping not to manipulate the location directly but the code above is not great either. I made the suggested change.